kernel/main.c

```diff
@@ -176,6 +173,12 @@ void stop_kernel(void)
 	printk("** Press Any Key to Reboot **\n");
 	any_key_to_reboot = 1;
 
+	/* stop and disable all interrupts! */
+	CLI();
+	for(n = 0; n < NR_IRQS; n++) {
+		disable_irq(n);
+	}
+
```

Comment:
I think that interrupts should be disabled **before** the last kernel messages.

Reply:
> I think that interrupts should be disabled **before** the last kernel messages.

Could you elaborate on why you would prefer that? My experience has been that if you are using a serial port console, this prevents the last messages from being displayed and pressing a key does nothing. This is the output I see with a serial port:

```
Welcome to FiwixOS 3.2 (ttyS0)
Kernel Release 1.5.0

(none) login: root
Password: 

Welcome to FiwixOS 3.2 for the i386 architecture.

    - Read the /root/README file to know about all the files installed.
    - Refer to the Installation CD-ROM for the source code and the license of
      every single package installed in this system.

[(root) ~]# shutdown -h 0

Broadcast message from root (ttyS0) (Fri Jan  5 17:23:47 2024):

The system is going down for system halt NOW!
INIT: Switching to runlevel: 0
INIT: Sending processes the TERM signal
INIT: Sending processes the KILL signal
Stopping atd:                                              [  OK  ]
Stopping crond:                                            [  OK  ]
WARNING: root 2 with count 10 (on dev 3,2)
WARNING: inode 2 with count 10 (on dev 3,2)
umount: /: Device or resource busy
Remounting remaining filesystems (if any) readonly
Halting the system...
```

When I move the IRQ disabling *after* printing the last kernel messages then I do see the kernel messages. It appears that the serial port I/O requires enabled interrupts. However, pressing a key does not reboot the kernel when using a serial port so it didn't fully solve the problem but I thought it was better.

Reply:
Yes, you're right. The last kernel messages does not appear on a serial console, but the cause is not really this although disabling interrupts **after** the messages seems fixes it.

The problem is in the serial console driver that is not managing correctly the serial port when the messages come from the kernel. You can test it using the magic SysRq key. If you press `Alt+SysRq+t' to display the current list of processes, you'll see that the list is incomplete (depending on the number of processes) on a serial console. I'm aware of this problem and I plan to fix it during the 1.5.0 version. Remember, I advice you some time ago, to increase the value of `NR_CB_QUEUE` in `include/fiwix/tty.h` if you wanted to see all that process list.

Meanwhile, leave the code as it is now. I mean, to disable interrupts **after** the kernel messages, so you'll notice if the kernel halted successfully, and so you'll have a better debugging experience.

Once I fixed the bug, I'll change the code to be **before** of the last kernel messages.

Reply:
By the way, I recommend you to use the latest FiwixOS 3.3 version.